packages/go/analysis/ad/esc4.go

```diff
@@ -206,3 +208,620 @@ func FetchPrincipalsWithWritePKIEnrollmentFlagOnCertTemplate(tx graph.Transactio
+}
+
+// composition: p12, p13, p14
+func findPathsToDomainThroughCertTemplateWithBothPKIFlags(
+	ctx context.Context,
+	db graph.Database,
+	startNode *graph.Node,
+	domainID graph.ID,
+	enterpriseCAs cardinality.Duplex[uint32],
+) (map[graph.ID][]*graph.PathSegment, cardinality.Duplex[uint32], error) {
+
+	var (
+		traversalInst = traversal.New(db, analysis.MaximumDatabaseParallelWorkers)
+		lock          = &sync.Mutex{}
+
+		certTemplateSegments = map[graph.ID][]*graph.PathSegment{}
+		certTemplates        = cardinality.NewBitmap32()
+	)
+
+	// p9: use the enterpriseCA nodes to gather the set of cert templates with an inbound `WritePKIEnrollmentFlag`
+	if err := traversalInst.BreadthFirst(ctx, traversal.Plan{
+		Root: startNode,
+		Driver: certTemplateWithPrivelegesToDomainTraversal([]graph.Kind{ad.WritePKIEnrollmentFlag}, domainID, enterpriseCAs).Do(
+			func(terminal *graph.PathSegment) error {
+
+				certTemplate := terminal.Search(
+					func(nextSegment *graph.PathSegment) bool {
+						return nextSegment.Node.Kinds.ContainsOneOf(ad.CertTemplate)
+					},
+				)
+
+				lock.Lock()
+				certTemplateSegments[certTemplate.ID] = append(certTemplateSegments[certTemplate.ID], terminal)
+				certTemplates.Add(certTemplate.ID.Uint32())
+				lock.Unlock()
+
+				return nil
+			}),
+	}); err != nil {
+		return certTemplateSegments, certTemplates, err
+	}
+
+	// p10: (reuse p4 logic): find cert templates that have an inbound `Enroll` OR `AllExtendedRights` edge
+	if err := traversalInst.BreadthFirst(ctx, traversal.Plan{
+		Root: startNode,
+		Driver: certTemplateWithEnrollmentRightsTraversal(certTemplates, nil).Do(
+			func(terminal *graph.PathSegment) error {
+
+				certTemplate := terminal.Search(
+					func(nextSegment *graph.PathSegment) bool {
+						return nextSegment.Node.Kinds.ContainsOneOf(ad.CertTemplate)
+					},
+				)
+
+				lock.Lock()
+				certTemplateSegments[certTemplate.ID] = append(certTemplateSegments[certTemplate.ID], terminal)
+				lock.Unlock()
+
+				return nil
+			}),
+	}); err != nil {
+		return certTemplateSegments, certTemplates, err
+	}
+
+	// p14: find cert templates with valid combination of properties that has an inbound `WritePKIName` edge
+	if err := traversalInst.BreadthFirst(ctx, traversal.Plan{
+		Root: startNode,
+		Driver: certTemplateWithPKINameFlagTraversal(certTemplates).Do(
+			func(terminal *graph.PathSegment) error {
+
+				certTemplate := terminal.Search(
+					func(nextSegment *graph.PathSegment) bool {
+						return nextSegment.Node.Kinds.ContainsOneOf(ad.CertTemplate)
+					},
+				)
+
+				lock.Lock()
+				certTemplateSegments[certTemplate.ID] = append(certTemplateSegments[certTemplate.ID], terminal)
+				lock.Unlock()
+
+				return nil
+			}),
+	}); err != nil {
+		return certTemplateSegments, certTemplates, err
+	}
+
+	return certTemplateSegments, certTemplates, nil
+}
+
+// composition: p2, p5, p8, p11, p15
+func findPathToDomainThroughEnterpriseCAsTrustedForNTAuth(
```

Comment:
Both `enterpriseCATraversal` and  `ntAuthStoreToDomainTraversal` start with an outbound with no max and min depth from the start node which can be expensive. Breaking these into different steps might involve starting at the eca's that are gathered in `enterpriseCATraversal` and continuing from there for the `ntAuthStoreToDomainTraversal` so as not to repeat the same expensive traversal initially.